libazureinit/src/provision/mod.rs

```
@@ -23,156 +24,116 @@ use crate::User;
-    /// provided with the [`User`].
-    pub fn password_provisioners(
-        mut self,
-        backend: impl Into<Vec<password::Provisioner>>,
-    ) -> Self {
-        self.password_backends = Some(backend.into());
-        self
-    }
-
-    /// Provision the host.
-    ///
-    /// Provisioning can fail if the host lacks the necessary tools. For example,
-    /// if there is no `useradd` command on the system's `PATH`, or if the command
-    /// returns an error, this will return an error. It does not attempt to undo
-    /// partial provisioning.
     #[instrument(skip_all)]
     pub fn provision(self) -> Result<(), Error> {
-        self.user_backends
-            .unwrap_or_else(|| user::Provisioner::iter().collect())
+        self.config
+            .hostname_provisioners
+            .backends
             .iter()
-            .find_map(|backend| {
-                backend
-                    .create(&self.user)
-                    .map_err(|e| {
-                        tracing::info!(
-                            error=?e,
-                            backend=?backend,
-                            resource="user",
-                            "Provisioning did not succeed"
-                        );
-                        e
-                    })
-                    .ok()
+            .find_map(|backend| match backend {
+                HostnameProvisioner::Hostnamectl => {
+                    hostname::Provisioner::Hostnamectl.set(&self.hostname).ok()
+                }
+                #[cfg(test)]
+                HostnameProvisioner::FakeHostnamectl => Some(()),
+            })
+            .ok_or(Error::NoHostnameProvisioner)?;
+
+        self.config
+            .user_provisioners
+            .backends
+            .iter()
+            .find_map(|backend| match backend {
+                UserProvisioner::Useradd => {
+                    user::Provisioner::Useradd.create(&self.user).ok()
```

Do we want to panic here if the user cannot be created?

---

I would say the same here as I did above -- was trying to mirror the original behavior and maintain predictability in behavior.  But, if people think panicking might make the error more immediately obvious, then can panic instead of returning ok_or(Error::NoUserProvisioner)!